src/main.rs

```diff
@@ -470,55 +526,189 @@ fn verify_event_attendance_and_get_event_name(spec: &CredentialSpec) -> Result<&
     }
 }
 
+// TODO: Create and manage admins instead of using canister controllers.
+async fn is_admin(id: Principal) -> bool {
+    is_controller(&id)
+}
+
 #[update]
 #[candid_method]
-fn register_early_adopter(
-    request: RegisterRequest,
-) -> Result<EarlyAdopterResponse, EarlyAdopterError> {
+async fn list_events() -> Result<ListEventsResponse, RegisterError> {
+    let user_id = caller();
+    let is_controller = is_admin(user_id).await;
```

Comment:
as you're wrapping controller-functionality behind admin-abstraction, maybe the variable should be called `is_admin` as well?  (otherwise the controller is popping up again)

Reply:
Oh, true, in Rust I can reuse the same name. Done!